flink-kubernetes-operator-autoscaler/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/config/AutoScalerOptions.java

```diff
@@ -98,6 +98,13 @@ private static ConfigOptions.OptionBuilder autoScalerConfig(String key) {
                     .withDescription(
                             "Max scale down factor. 1 means no limit on scale down, 0.6 means job can only be scaled down with 60% of the original parallelism.");
 
+    public static final ConfigOption<Double> MAX_SCALE_UP_FACTOR =
+            autoScalerConfig("scale-up.max-factor")
+                    .doubleType()
+                    .defaultValue(2.0)
```

Comment:
I wonder if we should set this higher by default, @mxm what do you think?

Reply:
In our production setup, we would definitely set this to a very high number, effectively disabling it. The reason is that we have pipelines which sit most idle (parallelism 1) but then receive huge traffic spikes which need quite high parallelisms (up to 500) to be able to keep up. The only way we limit the scale up is by setting a max parallelism.

Being able to scale up quickly is very important to keep up with the backlog size / growth. If you cap the scale up, you limit the effectiveness of the scaling. I get where it comes from but generally I think it is curing the symptoms and not the cause of these spikes.

Reply:
I would suggest to set the default to the maximum parallelism, to be able to scale up from 1 to the max configured parallelism.

Reply:
BTW, I'm quite interested in the scale, job type, business scenarios and of course all the problems encountered while adopting autoscaling in production. In my company, we encountered quite a few problems. It'll be good if we can have a discussion on flink-user mail-list sharing experience or even an online meetup would be great.

Reply:
For example, in my company:
1. Job input usu. has periodic pattern, sometimes spikes but smoothly change in most cases
2. Underlying resources are not elastic, resource competition between jobs is not rare
3. Job types are diverse with heavy dependence on various external services, scaling Flink alone is often ineffective
4. The scale is gigantic (over 40k jobs), we have to carefully limit the autoscaler's behavior to prevent resonant scaling behaviors from causing massive failures

Reply:
@X-czh Great comments! You are right that the autoscaler configuration is still quite important to handle load skew in big jobs. Concerning the constraints you listed, we definitely also fall under (1) although we are elastic to a large degree. It is definitely a good idea to configure the max parallelism even in an elastic cluster. 

Concerning (4) I'd be interested what you mean by massive failures? Jobs taking other jobs resources? One of the biggest concerns with the autoscaler currently is the lack of resource pre-allocation. Fortunately, this issue will soon be fixed once we start using Flink's Rescale API in the autoscaler instead of triggering a full redeploy every time.

Reply:
Hi @mxm, I'll expand a bit on (4). In case of a data center failure, we will go through a disaster recovery process, jobs will be migrated to another data center to recover. During the process,
- A huge backlog will be built up, and when these jobs recover, all jobs tend to scale up but the overall resource is limited.
- Many external services might not recover to in time or have to lower capacity for other higher priority services, this will cause Flink jobs replying on them process slower and tend to scale up (and very likely scale up a lot!) as well.
As a result, almost all the jobs will tend to scale up a lot during this process, and we need to limit the single-step scale-up behavior so that not too much resources are occupied by a small set of jobs before automatic scaling circuit breaker and human intervention to take action.

Reply:
But yes, our use case might be a bit extreme for most operator users. We tend to add tight constraints on scaling for the overall stability of our clusters while serving most "well-behaved" jobs well at the cost of limiting the elasticity of a single job. Here, "well-behaved" refer to the job input changing smoothly.